clippy_lints/src/lifetimes.rs

```diff
@@ -860,36 +856,108 @@ fn elision_suggestions(
         .collect::<Vec<_>>();
 
-    let mut suggestions = if elidable_lts.len() == explicit_params.len() {
+    if !elidable_lts
+        .iter()
+        .all(|lt| explicit_params.iter().any(|param| param.def_id == *lt))
+    {
+        return None;
+    }
+
+    let mut suggestions = if elidable_lts.is_empty() {
+        vec![]
+    } else if elidable_lts.len() == explicit_params.len() {
         // if all the params are elided remove the whole generic block
         //
         // fn x<'a>() {}
         //     ^^^^
         vec![(generics.span, String::new())]
     } else {
-        elidable_lts
-            .iter()
-            .map(|&id| {
-                let pos = explicit_params.iter().position(|param| param.def_id == id)?;
-                let param = explicit_params.get(pos)?;
-
-                let span = if let Some(next) = explicit_params.get(pos + 1) {
-                    // fn x<'prev, 'a, 'next>() {}
-                    //             ^^^^
-                    param.span.until(next.span)
+        match &explicit_params[..] {
+            // no params, nothing to elide
+            [] => unreachable!("handled by `elidable_lts.is_empty()`"),
+            [param] => {
+                if elidable_lts.contains(&param.def_id) {
+                    unreachable!("handled by `elidable_lts.len() == explicit_params.len()`")
                 } else {
-                    // `pos` should be at least 1 here, because the param in position 0 would either have a `next`
-                    // param or would have taken the `elidable_lts.len() == explicit_params.len()` branch.
-                    let prev = explicit_params.get(pos - 1)?;
-
-                    // fn x<'prev, 'a>() {}
-                    //           ^^^^
-                    param.span.with_lo(prev.span.hi())
+                    unreachable!("handled by `elidable_lts.is_empty()`")
+                }
+            },
+            [first, second] => {
+                // NOTE: we could theoretically remove this case for the next one,
+                // but the latter likely has quite a lot of overhead, so don't
+                match (
+                    elidable_lts.contains(&first.def_id),
+                    elidable_lts.contains(&second.def_id),
+                ) {
+                    (false, false) => unreachable!("handled by `elidable_lts.is_empty()`"),
+                    (true, true) => unreachable!("handled by `elidable_lts.len() == explicit_params.len()`"),
+
+                    // <'a, 'b> -> <'a, 'b>
+                    //      ^^        ^^^^
+                    (false, true) => vec![(second.span.with_lo(first.span.hi()), String::new())],
+                    // <'a, 'b> -> <'a, 'b>
+                    //  ^^          ^^^^
+                    (true, false) => vec![(first.span.until(second.span), String::new())],
+                }
+            },
+            // Time for the big guns
```

Comment:
I would remove this, as this is not on the hot path: it is called only when actually emitting the suggestion, and in this case we do not care much about some extra overhead.

```suggestion
```